File: server/core/src/main/java/io/eiren/util/OperatingSystem.java
package io.eiren.util;

import java.io.File;


public enum OperatingSystem {

	//@formatter:off
	LINUX("linux", new String[]{"linux", "unix"}),
	WINDOWS("windows", new String[]{"win"}),
	OSX("osx", new String[]{"mac"}),
	UNKNOWN("unknown", new String[0]);
	//@fomatter: on

	private final String[] aliases;
	public final String name;
	private static OperatingSystem currentPlatform;

	OperatingSystem(String name, String[] aliases) {
		this.aliases = aliases;
		this.name = name;
	}

	public static String getJavaExecutable(boolean forceConsole) {
		String separator = System.getProperty("file.separator");
		String path = System.getProperty("java.home") + separator + "bin" + separator;
		if(getCurrentPlatform() == WINDOWS) {
			if(!forceConsole && new File(path + "javaw.exe").isFile())
				return path + "javaw.exe";
			return path + "java.exe";
		}
		return path + "java";
	}

	public static OperatingSystem getCurrentPlatform() {
		if(currentPlatform != null)
			return currentPlatform;
		String osName = System.getProperty("os.name").toLowerCase();
		for(OperatingSystem os : values()) {
			for(String alias : os.aliases) {
				if(osName.contains(alias))
					return currentPlatform = os;
			}
		}
		return UNKNOWN;
	}

	public static String getTempDirectory() {
		if(currentPlatform == LINUX) {
			String tmp = System.getenv("XDG_RUNTIME_DIR");
			if(tmp != null) return tmp;
		}
		return System.getProperty("java.io.tmpdir");
	}
}

File: server/core/src/main/java/io/eiren/util/OperatingSystem.kt
package io.eiren.util

import java.io.File
import java.nio.file.Path
import java.util.*
import kotlin.io.path.Path

enum class OperatingSystem(
	val descriptor: String,
	private val aliases: Array<String>,
) {
	LINUX("linux", arrayOf("linux", "unix")),
	WINDOWS("windows", arrayOf("win")),
	OSX("osx", arrayOf("mac")),
	UNKNOWN("unknown", arrayOf()),
	;

	companion object {
		val currentPlatform: OperatingSystem by lazy {
			val osName = System.getProperty("os.name").lowercase(Locale.getDefault())
			entries.find { os -> os.aliases.any { alias -> osName.contains(alias) } }?.let {
				return@lazy it
			}
			UNKNOWN
		}
		fun getJavaExecutable(forceConsole: Boolean): String {
			val separator = System.getProperty("file.separator")
			val path = System.getProperty("java.home") + separator + "bin" + separator
			return if (currentPlatform == WINDOWS) {
				if (!forceConsole && File(path + "javaw.exe").isFile) path + "javaw.exe" else path + "java.exe"
			} else {
				path + "java"
			}
		}

		val socketDirectory: String
			get() {
				var dir = System.getenv("SLIMEVR_SOCKET_DIR")
				if (dir != null) return dir
				if (currentPlatform == LINUX) {
					dir = System.getenv("XDG_RUNTIME_DIR")

					// add /app/$FLATPAK_ID if running in flatpak
					// see https://docs.flatpak.org/en/latest/sandbox-permissions.html
					val flatpak_id = System.getenv("FLATPAK_ID")
					if (flatpak_id != null) dir += "/app/" + flatpak_id

					if (dir != null) return dir
				}
				return System.getProperty("java.io.tmpdir")
			}

		fun resolveConfigDirectory(identifier: String): Path? = when (currentPlatform) {
			LINUX -> System.getenv("XDG_CONFIG_HOME")?.let { Path(it, identifier) }
				?: System.getenv("HOME")?.let { Path(it, ".config", identifier) }

			WINDOWS -> System.getenv("AppData")?.let { Path(it, identifier) }

			OSX -> System.getenv("HOME")?.let { Path(it, "Library", "Application Support", identifier) }

			UNKNOWN -> null
		}

		fun resolveLogDirectory(identifier: String): Path? = when (currentPlatform) {
			LINUX -> System.getenv("XDG_DATA_HOME")?.let { Path(it, identifier, "logs") }
				?: System.getenv("HOME")?.let { Path(it, ".local", "share", identifier, "logs") }

			WINDOWS -> System.getenv("AppData")?.let { Path(it, identifier, "logs") }

			OSX -> System.getenv("HOME")?.let { Path(it, "Library", "Logs", identifier) }

			UNKNOWN -> null
		}
	}
}
